### PR TITLE
[WIP] Use SymEngine in sympy/physics

### PR DESCRIPTION
**Pull request by ShikharJ**

### CODE REVIEW

**Comment by gxyd** [comment]
@ShikharJ can you help me understand the use of all the PR's adding a 'backend' as SymEngine. And the plan for getting all the PR's related to those changes in SymPy, a brief explanation would suffice. So that we can get the PR's merged in a well defined manner.

**Comment by ShikharJ** [comment]
The idea is that while calling a method from a particular directory (take the physics module as an example), through the use of flags (`USE_SYMENGINE=1`), the underlying routines used by the particular method (like `sqrt`, `S` and `factorial` for `R_nl`) should be sourced from `SymEngine` for speed benefits. This is to be done using the `sympy.core.backend` file, which by default switches to `SymPy`'s routines when the flag is not set (keeping the changes required minimal). Also, currently only those methods are made available in `backend.py` which are implemented in `SymEngine` as well as `SymPy`, rest are imported from within `SymPy` only.

**Comment by ShikharJ** [comment]
Currently, for the maximum benefit, we'll have to implement the lacking functions in `SymEngine` and `SymEngine.py` (only a few), try porting them in `SymPy` and then weeding out the inconsistencies (which is the bigger challenge as the SymPy library is huge and interconnected). For example, many routines in `SymEngine` do not take extra keywords (`expand` has no provision for `complex` keyword in `SymEngine` but it is used in `SymPy` and hence it is an inconsistency that has to be cleared out).

**Comment by gxyd** [comment]
I understand to some extent, thanks for the clarification. If these pull requests are ready, then can we polite ping someone who can review these pull requests?

**Comment by ShikharJ** [comment]
I'd personally prefer if we've completely covered a specific module or sub-module before merging. I'll start off working on them right away. Though some can be merged in their present state, but then we wouldn't be able to extract every last benefit of `SymEngine`'s performance as pointed out by @isuruf. So, I'll leave it to him to make the call.

File: sympy/physics/tests/test_physics_matrices.py [hunk 1]
```diff
@@ -70,8 +69,10 @@ def test_Dirac():
 
 def test_mdft():
     assert mdft(1) == Matrix([[1]])
-    assert mdft(2) == 1/sqrt(2)*Matrix([[1,1],[1,-1]])
+    assert mdft(2) == 1/sqrt(2)*Matrix([[1,1],[1,exp(-I*pi)]])
```
Comment (certik) [inline sympy/physics/tests/test_physics_matrices.py]: Isn't `exp(-I*pi)` equal to `-1`? Why is this change needed?

**Review by certik — COMMENTED** [review]
Otherwise it looks good.

File: sympy/physics/tests/test_physics_matrices.py (hunk 1 above)
Comment (ShikharJ) [reply sympy/physics/tests/test_physics_matrices.py]: @certik If I'm not mistaken, such expressions are not simplified in `SymEngine`.

Comment (certik) [reply sympy/physics/tests/test_physics_matrices.py]: I see. I think that's an omission, I just created an issue for it: https://github.com/symengine/symengine/issues/1330

**Comment by ShikharJ** [comment]
Same as [here](https://github.com/sympy/sympy/pull/13214#issuecomment-533908620).
